cdpinger/man/lbmount.1
.TH "lbmount" "1" 
.SH "NAME" 
lbmount \(em mount ltspfs based filesystems to a user owned subdir in /media 
.SH "SYNOPSIS" 
.PP 
\fBlbmount\fR 
.SH "DESCRIPTION" 
.PP 
This manual page documents briefly the 
\fBlbmount\fR       command. 
 
.PP 
\fBlbmount\fR is a suid wrapper that creates mountpoints in /media and bind mounts ltspfs mounted devices to there 
\fBlbmount\fR 
.SH "OPTIONS" 
.PP 
Usage: lbmount [\-\-umount] mediadir 
 
.SH "SEE ALSO" 
.PP 
ltspfs (1). ltspfsmounter (1). 
.SH "AUTHOR" 
.PP 
This manual page was written by Oliver Grawert ogra@ubuntu.com for 
the \fBDebian\fP system (but may be used by others).  Permission is 
granted to copy, distribute and/or modify this document under 
the terms of the GNU General Public License, Version 2 any  
later version published by the Free Software Foundation. 
 
.PP 
On Debian systems, the complete text of the GNU General Public 
License can be found in /usr/share/common-licenses/GPL. 
 
.\" created by instant / docbook-to-man, Tue 29 Aug 2006, 11:56 


cdpinger/ChangeLog
  157 Vagrant Cascadian 2012-02-24
      Add symlinks for Debian/Ubuntu for ltspfsd.rules creation in
      init-ltsp.d.

  156 Vagrant Cascadian 2012-02-24
      Use init-ltsp.d hook for installing udev rules at boot.

  155 Vagrant Cascadian 2012-02-24
      Rename udev rules to ltspfsd.rules, and move them into
      /usr/share/ltspfs/udev, which allows for installing rules at boot.

  154 vagrant@freegeek.org	2011-10-28 {ltspfs-1.0}
      version 1.0

  153 vagrant@freegeek.org	2011-10-28 [merge]
      merge patch from Scott Balneaves to fix 64 bit issues.

  152 Stéphane Graber	2011-08-11 {ltspfs-0.9}
      version 0.9

  151 Stéphane Graber	2011-07-22
      Launchpad: no change commit to trigger a test rebuild

  150 Stéphane Graber	2011-07-07
      ID_FS_LABEL is now usually 'safe'. Patch from gcc (LP: #690969)

  149 Warren Togami	2011-06-16 {ltspfs-0.8}
      version 0.8

  148 Vagrant Cascadian 2011-02-05
      cdpinger: add support for UDF formatted CDs and DVDs.

  147 Stéphane Graber	2010-08-08 {ltspfs-0.7}
      Release 0.7

  146 Vagrant Cascadian 2010-06-25
      init with "ltspfs" rather than simply "test".

  145 Vagrant Cascadian 2010-06-25
      Add example ltspfs mounted hook to issue notifications for
      desktops that do not display icons. Thanks to Petter Reinholdtsen.
      http://bugs.debian.org/575031

  144 Gideon Romm	2010-05-11
      Remove commented-out code.

  143 Gideon Romm	2010-05-11
      Fix cdrom handling.

  142 Stéphane Graber	2010-02-17 {ltspfs-0.6}
      version 0.6

  141 Stéphane Graber	2010-02-17
      Try umounting with -l to avoid some issues

  140 Scott Balneaves	2010-02-17
      Change owner of dir in /media to user, with group of root

  139 Scott Balneaves	2010-02-17
      Top level media directory being created mode 750.  Changed to 700

  138 Gideon Romm	2010-02-10
      Be careful to actually have a pid before executing kill

  137 Gideon Romm	2010-02-10
      Do not sed -i a file that we are reading in from stdin

  136 Vagrant Cascadian 2010-02-10
      allow CDs to have an empty filesystem, as they're likely to have
      a device but no disk inserted. this allows cdpinger to run again.

  135 Vagrant Cascadian 2010-02-10
      cdpinger: report syslog messages for "disk detected" and "tray
      open" messages only when the state changes so they are not reported
      every 3 seconds.

  134 Gideon Romm	2010-02-10
      Fix MODE setting to include add_disc annd remove_disc

  133 Gideon Romm	2010-02-10
      Pass DEVICENAME to add_cdrom_device and remove_cdrom_device
      functions

  132 Gideon Romm	2010-02-10
      Make sure ltspfsd gets launched

  131 Gideon Romm	2010-02-10
      Refactor ltspfs_entry and cdpinger code to better handle CDROMs

  130 Gideon Romm	2010-02-10
      Remove cdpinger pid file when cdrom is removed, since apparently
      cdpinger will commit suicide on its own upon device removal and
      we would like cdpinger to start up again upon reinsertion of the
      cdrom drive.

  129 Gideon Romm	2010-02-09
      Try to avoid multiple parallel calls to pgrep by writing a pid
      file to check first. If the pid file exists, it shell should not
      need to use pgrep in the conditional

  128 Gideon Romm	2010-02-01
      Make delayed-mounter work for LDM localapps

  127 Scott Balneaves	2010-01-30
      Added syslogging.

  126 Gideon Romm	2010-01-29
       Some more optimizations to use shell primitives where possible

  125 Vagrant Cascadian 2010-01-29
      fix call_cdpinger function to use exact matching so that "cdrom"
      doesn't match "cdrom1". thanks gadi!

  124 Gideon Romm	2010-01-27
      Fix up noauth stuff and make local devices work for localapps,
      rdesktop, and kiosk scripts.

  123 Scott Balneaves	2010-01-27
      Changed noauth mode: ltspfs now recognizes LTSPFS_NOAUTH env var,
      and if set, skips auth.  ltspfsd now skips receiving auth packet
      if noauth set.

  122 Stéphane Graber	2010-01-13 {ltspfs-0.5.14}
      release ltspfs 0.5.14

  121 Stéphane Graber	2010-01-13
      Fix cases where the user used for login doesn't match the case of
      the remote user

  120 Scott Balneaves	2010-01-12
      Files are now user-owned only

  119 Marc Gariepy	2009-12-15
      fixing local lowercase username (ldap ltsp-trunk -r 1265)

  118 Vagrant Cascadian 2009-07-17 {ltspfs-0.5.13}
      release ltspfs 0.5.13

  117 Vagrant Cascadian 2009-07-17
      drop delayedmounter from xinitrc.d, as udev scripts should be
      sufficient now.

  116 Vagrant Cascadian 2009-07-06
      relocate delayed-mounter ldm hooks so they run after the user
      is added to /etc/passwd, and before the user is removed from
      /etc/passwd.

  115 Vagrant Cascadian 2009-07-06
      - always run as root, so mounts appear in /media/root - only run
      mount for localapps if user's home dir is mounted via sshfs

  114 Vagrant Cascadian 2009-07-06
      check for ldm-xauth in /var/run, and support newer ldm, with
      Xauthority in a sub-dir.

  113 Vagrant Cascadian 2009-06-21
      fix spelling of "Received" in ltspfsd

  112 Stéphane Graber	2009-06-03 {ltspfs-0.5.12}
      version 0.5.12

  111 Francis Giraldeau 2009-05-27
      remove ltspfs_fstab entry before unmounting device,
      to prevent other processes from mounting it.
      https://bugs.launchpad.net/ubuntu/+source/ltspfsd/+bug/378495

  110 Vagrant Cascadian 2009-04-03 {ltspfs-0.5.11}
      version 0.5.11

  109 Vagrant Cascadian 2009-04-03
      simplify code to handle when mountpoint is already mounted

  108 Gideon Romm	2009-03-11
      Make sure delayed_mounter sets USER=root for local ltspfsmounting

  107 Gideon Romm	2009-03-10
      Do not remove ltspfs mount on "already mounted" errors, as these
      errors may occur silently with multiple ltspfsmounters

  106 Stéphane Graber	2009-03-09 {ltspfs-0.5.10}
      version 0.5.10

  105 Stéphane Graber	2009-03-08

cdpinger/src/lbmount.c
/*
 * lbmount.c: - setuid wrapper around a bindmount to allow ltsp localdevices
 *              to appear under /media
 *
 * Copyright Scott Balneaves, sbalneav@ltsp.org, 2006, 2007
 *           Martin Pitt, 2006
 *           Vagrant Cascadian, vagrant@freegeek.org, 2008
 *           Warren Togami, wtogami@redhat.com, 2008

 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, you can find it on the World Wide
 * Web at http://www.gnu.org/copyleft/gpl.html, or write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 * MA 02110-1301, USA.

 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <pwd.h>
#include <mntent.h>
#include <limits.h>
#include <errno.h>
#include <getopt.h>

static uid_t uidReal;           /* Users real userid */

/*
 * Static definitions of some parts of the mount.  The program only gets passed
 * the name of the mounted device (i.e. the variable "mountpoint", and won't
 * accept any pathing on this variable.  Then, the path is built dynamically.
 * So, a bindmount from /tmp/.username-ltspfs/devicename to 
 * /media/username/devicename is made.  Below are the "static" pieces of the
 * mount strings, used later to dynamically build the full paths based on 
 * the username and mountpoint supplied. 
 */

static char *mediadir = "/media";
static char *ltspfsdir1 = "/tmp/.";
static char *ltspfsdir2 = "-ltspfs/";
static char *mountprog = "/bin/mount";  /* system mount program */
static char *umountprog = "/bin/umount";        /* system umount program */

void usage(char *progname)
{
    fprintf(stderr, "usage: %s [--umount] mediadir\n", progname);
    exit(0);
}

/*
 * is_mounted: checks a path to see if it exists as a mountpoint in /proc/mounts
 */

int is_mounted(char *path)
{
    FILE *f;
    struct mntent *entry;

    if (!(f = fopen("/proc/mounts", "r"))) {
        perror("Error: could not open /proc/mounts");
        exit(1);
    }

    while ((entry = getmntent(f)) != NULL) {
        if (!strcmp(path, entry->mnt_dir)) {
            /* OK, here's where things get a little odd.  We want to make sure
             * that the mount is owned by the user.  Normally, we'd do a stat
             * call here, but because we're a setuid program, and fuse by
             * default doesn't even allow ROOT to access users mounts, a stat
             * call will fail.  So, what we'll do is look at the "user_id=xxxx"
             * option in the mount tab, and see if it's the same as our own.
             * Clever.... veeeeery clever */
            char *pos;
            pos = strstr(entry->mnt_opts, "user_id");
            if (pos) {
                uid_t mounted_uid;
                pos += 8;                        /* skip over user_id= */
                sscanf(pos, "%d", &mounted_uid);
                if (mounted_uid == uidReal) {
                    endmntent(f);
                    return 1;
                }
            } else {
                fprintf(stderr,
                        "Error: mountpoint doesn't have user_id= in options\n");
            }
        }
    }

    fprintf(stderr, "Error: %s is not mounted\n", path);
    endmntent(f);
    return 0;
}

void mkdir_safe(char *dir)
{
    if (mkdir(dir, 0700)) {
        if (errno == EEXIST) {
            /* OK, we've been told that this already exists.  Make sure
             * nothing's mounted on top of it, so we aren't doing something
             * nasty */
            if (is_mounted(dir)) {
                fprintf(stderr,
                        "Error: can't bindmount under %s: already mounted\n",
                        dir);
                exit(1);
            }
        } else {
            perror("Unable to mkdir() in /media");
            exit(1);
        }
    }
    if (chown (dir, uidReal, 0)) {
	perror ("Unable to change owner of directory");
	rmdir (dir);
	exit (1);
    }
}

/*
 * domount: actually bindmounts path1 onto path2.
 */

int root_mounter(const char *path1, const char *path2)
{
    int status;
    pid_t child;
    char *program;
    char *null_env[] = { NULL };

    child = fork();

    if (child == 0) {
        if (setreuid(0, -1)) {
            /* Couldn't become root */
            perror("Couldn't obtain root privs");
            exit(1);
        }
        /* Statically build command line to prevent monkey business */
        if (path2)
            execle(mountprog, mountprog, "--move", path1, path2, NULL,
                   null_env);
        else
            execle(umountprog, umountprog, "-l", path1, NULL, null_env);
        perror("Error: execl() returned");
        exit(1);                                 /* exec should never return */
    } else if (child > 0) {
        if (waitpid(child, &status, 0) < 0) {
            perror("Error: wait() call failed");
            exit(1);
        }
    } else if (child < 0) {
        perror("Error: fork() failed");
        exit(1);
    }

    if (!WIFEXITED(status)) {
        fprintf(stderr, "Error: execle() returned no status");
        exit(1);
    }

    return WEXITSTATUS(status);
}


/*
 * mainline
 */

int main(int argc, char **argv)
{
    int umount = 0;
    struct passwd *pwent;
    char *mountpoint = NULL;    /* command line supplied media name */
    char mediamount[PATH_MAX];  /* fully pathed mountpoint in /media */
    char ltspfsmount[PATH_MAX]; /* fully pathed ltspfs mount in /tmp */

    int option;
    static struct option long_opts[] = {
        {"help", 0, NULL, 'h'},
        {"umount", 0, NULL, 'u'},
        {NULL, 0, NULL, 0}
    };

    /* Save our real userid */
    uidReal = getuid();

    /* Is the setuid bit set? */
    if (geteuid()) {
        perror("Error: progarm not installed setuid root");
        exit(1);
    }

    pwent = getpwuid(uidReal);
    if (!pwent) {
        perror("uid doesn't appear to have a valid passwd entry");
        exit(1);
    }

    /* Command line handling */
    do {
        option = getopt_long(argc, argv, "hu", long_opts, NULL);
        switch (option) {
        case -1:
            break;
        case 'h':
            usage(argv[0]);
            return 0;
        case 'u':
            umount = 1;
            break;
        default:
            fprintf(stderr, "%c: unknown option\n", option);
            exit(1);
        }
    }
    while (option != -1);

    /* get the optional mountpoint */
    if (optind < argc) {
        mountpoint = strdup(argv[optind]);
        if (!mountpoint) {
            fprintf(stderr, "Error: couldn't get mountpoint");
            exit(1);
        }
    } else {
        fprintf(stderr, "Error: no mountpoint supplied\n");
        exit(1);
    }

    /* security: mountpoint cannot have any /'s in it, or be too long */
    if (strlen(mountpoint) > 80 || index(mountpoint, '/')) {
        fprintf(stderr, "mountpoint name invalid.\n");
        exit(1);
    }

    /* security: make sure pwd->pw_name doesn't have a '/' in it */
    if (index(pwent->pw_name, '/')) {
        fprintf(stderr, "username invalid.\n");
        exit(1);
    }

    /*
     * Build our ltspfs mountpoint string, and check and see if it exists.
     */

    snprintf(ltspfsmount, sizeof(ltspfsmount), "%s%s%s%s",
             ltspfsdir1, pwent->pw_name, ltspfsdir2, mountpoint);

    if (!umount) {
        if (!is_mounted(ltspfsmount))
            exit(1);

        /* OK, name's a normal size, and looks valid. Begin creating the media
         * mount point. First, we need to create /media/uid */

        snprintf(mediamount, sizeof(mediamount), "%s/%s", mediadir,
                 pwent->pw_name);

        mkdir_safe(mediamount);

        /* Now, create the media dir underneath the uid dir */

        snprintf(mediamount, sizeof(mediamount), "%s/%s/%s", mediadir,
                 pwent->pw_name, mountpoint);

        mkdir_safe(mediamount);

        return root_mounter(ltspfsmount, mediamount);
    } else {
        /* umount */

        snprintf(mediamount, sizeof(mediamount), "%s/%s/%s", mediadir,
                 pwent->pw_name, mountpoint);

        if (is_mounted(mediamount))
            root_mounter(mediamount, NULL);
        else {
            fprintf(stderr, "Error: %s unmountable", mediamount);
            exit(1);
        }

        if (rmdir(mediamount)) {
            perror("Unable to rmdir() in /media");
            exit(1);
        }

        snprintf(mediamount, sizeof(mediamount), "%s/%s", mediadir,
                 pwent->pw_name);

        if (rmdir(mediamount)) {
            /* If we get an ENOTEMPTY, we can pass it by, as there's other
             * mounts still there. */
            if (errno != ENOTEMPTY) {
                perror("Unable to rmdir() users dir in /media");
                exit(1);
            }
        }

        exit(0);
    }
}
